### PR TITLE
Added support for check_same_thread

### PR DESCRIPTION
**Pull request by pkerpedjiev**
This pull request addresses issue https://github.com/biopython/biopython/issues/2234

- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
Is there a downside to doing ``check_same_thread=False`` as the default? i.e. Don't change the ``Bio.SeqIO`` or ``Bio.SearchIO`` (etc) interface at all.

CC @bow

**Comment by pkerpedjiev** [comment]
Afaict, it just puts the onus on the user to make sure write operations don't conflict with each other. In the case of sequence indexes that shouldn't be an issue since the index creation can't be done in a multithreaded fashion. 

> By default, check_same_thread is True and only the creating thread may use the connection. If set False, the returned connection may be shared across multiple threads. When using multiple threads with the same connection writing operations should be serialized by the user to avoid data corruption.

If you're ok with making check_same_thread=false the default then I'm happy to modify the PR.

**Comment by peterjc** [comment]
For our read only interface in ``Bio/File.py`` class ``_SQLiteManySeqFilesDict``, I am inclined to just change the internal connection setup:

```python
        con = _sqlite.connect(index_filename,
                              check_same_thread=self._check_same_thread)
```

i.e. Don't change our user facing API at all.

Does that seem sensible?

**Comment by pkerpedjiev** [comment]
> Does that seem sensible?

Yes. I'll modify the PR.

**Comment by peterjc** [comment]
You probably understood what I meant, but I think I should have written:

```python
con = _sqlite.connect(index_filename, check_same_thread=False)
```

**Comment by pkerpedjiev-zymergen** [comment]
Yup, I figured you just took the line from the PR to point where you meant for that to go.

File: Tests/test_SeqIO_index.py [hunk 1]
```diff
@@ -111,6 +112,20 @@ def test_old(self):
             self.assertEqual(54, len(d))
             self.assertRaises(FileNotFoundError, d.get_raw, "alpha")
 
+        def test_old_check_same_thread(self):
+            """Setting check_same_thread to False doesn't raise an exception"""
```
Comment (peterjc) [inline Tests/test_SeqIO_index.py]: Needs a trailing full stop,

```
$ flake8
./Tests/test_SeqIO_index.py:116:1: D400 First line should end with a period
```

Comment (peterjc) [reply Tests/test_SeqIO_index.py]: Strange - you're already in the top level Biopython folder so https://gitlab.com/pycqa/flake8/issues/517 shouldn't be the problem. Maybe the version of flake8 is important...

Comment (peterjc) [reply Tests/test_SeqIO_index.py]: As to the task at hand, we currenlty tell flake8 to ignore the following under ``Tests/``:

F401,F841,D101,D102,D103,B009,B010,B011,C812

That is on top of the things we ignore everywhere like E501 - hopefully the ``.flake8`` file comments are clear.

i.e. D400 is the only one you should be seeing here.

**Comment by peterjc** [comment]
Any comments @bow?

**Comment by peterjc** [comment]
See also fc8ac0ea0048e152e409d229fa85aee205c95c01 as I missed the use of single quotes when merging this (wasn't enforced when the PR was tested).

**Pull request merged**